### ISSUE TITLE
Solving Sync/Async dispatch tension with a global re-entrant lock

### ISSUE THREAD

**Issue by yelouafi** [yelouafi/redux-saga#235]
## Problem

To handle the `put` effect (i.e. when a saga does `yield put(action)`) we actually have 2 possibilities

#### 1. Execute the dispatch synchronously (immediately)

```javascript
runPutEffect(action, next) {
  next( dispatch(action) )
}
```

#### 2. Execute the dispatch asynchronously (scheduling in a microtask using `Promise.then`

```javascript
runPutEffect(action, next) {
  Promise.resolve().then(() => next( dispatch(action) ))
}
```

Both approaches solves issues of each other. But not their proper issues

#### 1. The first one is *startup saga actions*. Actions which are fired by a Saga immediately after it is started.

Let's see a simple example

```javascript
function* saga() {
  yield put(action)
}

const store = createStore(
  reducer, 
  applyMiddleware(
    createSagaMiddleware(saga)
  )
)
```

The issue is that `createSagaMiddleware(saga)` will starts the Saga while still in the `applyMiddleware` phase, i.e. before the `applyMiddleware` returns. This will result in the middleware taking and executing the `put(action)` Effect. But we can't actually dispatches actions during middleware setup (see this issue https://github.com/reactjs/redux/issues/1240).

The immediate solution is to delay the startup of sagas until the store creation completes, but Redux do not actually have a lifecycle API for middleware (like with react) so the only way left is to start the Saga asynchronously using microtasks

```javascript
sagaMiddleware = (saga) => ({getState, dispatch}) => {
  Promise.resolve().then(() => startSaga(saga))
  return next => action ...
}
```

Now the saga startup will be delayed, and when the promise executes its then callback, the store creation will be already completed. so the dispatch can be executed without problem. But imagine now that the first instruction of `saga` was a `take`

```javascript
function* saga() {
  yield take(actionType)
}
```

If we delay the saga startup, then any actions dispatched synchronously after the store creation (e.g. by components is `componentDidMount`) will be missed by the saga, because it hasn't yet started and thus hasn't yet yielded the `take` effect. So we now have an issue with *startup store actions*.

We can go further and try to save all actions dispatched before saga startup in a temporary queue then delivers them once the saga starts on the next microtask but this will create other issues; plus there is another issue with synchronous dispatches

#### 2. The second issue is related to nested dispatches

it can occur for example when you have 2 sagas sending actions to each other. Here is an example (this is not specific to Sagas, it can occur for ex. for 2 middlewares communicating with each other)

```javascript
function* fnA() {
    yield take('msg-1'))
    yield put({type: 'ack-1'})
    yield take('msg-2'))
    yield put({type: 'ack-2'})
  }

  function* fnB() {
    yield put({type: 'msg-1'})
    yield take('ack-1')
    yield put({type: 'msg-2'})
    yield take('ack-2')
  }
```

Above `fnA` starts by waiting for a `msg-1` action. Assuming `fnA` starts the first, when `fnB` dispatches `msg-1` then `fnA` will take it and immediately dispatches an `ack-1` action which `fnB` is supposed to take.

The issue is that since `take` and `put` are both resolved synchronously, executing fnB's `put({type: 'msg-1'})` will cause the fnA's `take('msg-1')` to resolve withing the same stack frame.

```
fnB.............................fnA.........................

.................................take('msg-1'))..........  waiting  
put({type: 'msg-1'}).......resolve the take......  
.................................put({type: 'ack-1'})..... fnB put hasn't yet returned  
take('ack-1').............................................. will miss fnA put  
```

For a real example see #198

So seems synchronous dispatching is the wrong choice, What about asynchronous dispatching

### Issues of Asynchronous dispatch

Async. dispatching solves the 2 above problems by wrapping each saga's `put` inside a `Promise.then`. 

For the first issue, since the `put` is delayed to the next microtask, the dispatch won't occur until the store creation has completed. And since the saga startup itself isn't delayed it won't miss any action dispatched on the startup from outside (Component's actions dispatched in componentDidMount for example).

For the second issue, it's not obvious but it's also solved by async dispatches, Nested puts will be scheduled using Promise.then, which will cause them to be queued one after another. In the example above the 2nd put of `fnA` will be scheduled after `fnB` completes the 1st put and resume to `yield take('ack-1')`. So when the delayed `put({type: 'ack-1'})`  executes, `fnB` is already ready to take it.

This is the solution implemented actually in redux-saga; but the asynchronous dispatch has also an issue: since the put is executed asynchronously, the saga will remain blocked until the put resolves on the next microtask. For example

```javascript
function* saga() {
  yield take('a')
  yield put({type: 'a', payload: 'ack-1'})
  yield take('a')
  yield put({type: 'a', payload: 'ack-2'})
}

 store.dispatch({type: 'a', payload: 1})
 store.dispatch({type: 'a', payload: 2})
```

`saga` will take the 1st action and immediately : 1- put the `ack-1` action then 2- take another `a` action. But since the put will cause the saga to block on the current task, the 2nd ` store.dispatch({type: 'a', payload: 2})` will be fired while still the Saga is blocked (this is also discussed in https://github.com/yelouafi/redux-saga/issues/198#issuecomment-197905849).

### To summarize

- Synchronous dispatches do not block the Saga and thus no action is missed. But have issue with 
   1- startup saga actions (dispatched in the applyMiddleware phase) and   
   2- nested dispatches

- Asynchronous dispatches solve the issue of startup saga actions, and nested dispatches but cause the Saga to block during the dispatch and potentially miss actions (although the use case for this are pretty rare compared to the issues caused by (1)).

## Proposed solution 

so we have the following requirements

1- Sagas have to be started right away in order to be able to take actions dispatched at startup by external code and components   
2- Sagas must not block on puts to not miss any action  
3- We must take care of the nested dispatches issue  
4- Sagas must be able to dispatch actions at the startup

We have already (1) in place.
 
The proposed solution involves 2 changes

#### 1- Use a re-entrant global lock for all `put` effects

The re-entrant lock ([implementation](https://github.com/yelouafi/redux-saga/blob/master/src/internal/asap.js)) behaves similarly to Promise.then: The put effects are now resolved using a new scheduler `asap`

```javascript
runPutEffect(action, next) {
 asap(() => next( dispatch(action) ))
}
```

Anytime you call `asap(task)`, the scheduler will first check is there is another task already running: if not then it'll start the task immediately, if there are tasks running, it'll add the task to a task queue, the task will be executed after all previously added tasks are executed.

The difference from `Promise.then` is that:

- The Saga wont block for the whole current task (stack frame) but only after all pending dispatches completes. So the Saga will always resume before any outside dispatch occur. This solve the requirement (2)

- Since all puts passes through the `asap` lock, nested dispatches will be flattened and ordered just like we have actually with Promise.then (but without the blocking issue). So this solve (3)

to solve the remaining requirement (4) w'll need to introduce a breaking change :

#### 2- Change the middleware startup API by requiring the Saga to be started explicitly using [`middleware.run`](http://yelouafi.github.io/redux-saga/docs/api/index.html#middlewarerunsaga-args)

until now w're using this core to start the Sagas

```javascript
import { createStore, applyMiddleware } from 'redux'
import createSagaMiddleware from 'redux-saga'
import sagas from './sagas'
import rootReducer from './reducers'

const store = createStore(
  rootReducer,
  applyMiddleware(
    createSagaMiddleware(...sagas)
  )
)
```

With the new approach, all sagas are required to be started after store creation

```javascript
import { createStore, applyMiddleware } from 'redux'
import createSagaMiddleware from 'redux-saga'
import sagas from './sagas'
import rootReducer from './reducers'

const sagaRunner = createSagaMiddleware(/* options */)

const store = createStore(
  rootReducer,
  applyMiddleware(sagaRunner)
)

const tasks = sagas.map( saga => sagaRunner.run(saga) )
```

W're now sure that all puts from sagas are correctly dispatched since all sagas are started after store creation completes.

Another benefit is that you can now have a reference to tasks representing the started sagas. With the previous API, there was no way to get references to tasks started inside applyMiddleware.

The middleware setup is usually located in 1 or 2 places so (I think) it's no big deal to change the startup API, compared to the issues it solves (unless I miss something)

Actually the code is already in the master branch, but not released yet. There is also [the test suite](https://github.com/yelouafi/redux-saga/blob/master/test/proc/takeSync.js) documenting the above requirements,and any more tests for eventual edge cases is welcome.

What do you think of this:

- Are there any potential issue with the re-entrant lock solution?  
- Are there any potential issue with changing the startup API?
- Are there other less-complex solutions to the 4 requirements above?

**Comment by cherta** [comment]
Im not an expert @yelouafi but I won't mind change the way my apps instantiate the sagas because, as you said, I'm doing it only once per app. Nice explanation BTW.

**Comment by ctrlplusb** [comment]
Perhaps not helpful in this context, or may already be extremely obvious, but hopefully of interest to some:

In terms of issues with synchronous execution of startup sagas my technique is to fire an action (e.g. "APP_INIT") directly after my redux store has been configured.  My run once saga picks it up no problem and the app state is as expected before any component mounting occurs.

**Comment by slorber** [comment]
Like @ctrlplusb I always start my app with a startup events so I don't have much this kind of problem.



@yelouafi I had the same problem 2 years ago, when I built a hacky saga system on top of my [AtomReact stores](https://github.com/stample/atom-react/blob/master/src/atomReactStore.js#L49)
As you can see I choose the async path based on setTimeout :) but this was really annoying me because of the pros and cons you mention I didn't really know what to choose :D


I implemented something that looks a bit similar to your reentrant lock at some point ([command queuing](https://github.com/stample/atom-react/blob/master/src/atomReactContext.js#L227). Don't even really remember how it worked but it was quite hacky :)

I guess the idea is that the sagas should be able to become "stable" before receiving any new dispatch from outside the sagas right? If yes i'm fine with that.


However a nice default setting is cool, but I think it would be nice to give options in the put effect so that developers can decide themselves if they need sync/async/default for their puts

**Comment by yelouafi** [comment]
@ctrlplusb 
>In terms of issues with synchronous execution of startup sagas my technique is to fire an action (e.g. "APP_INIT") directly after my redux store has been configured

Yes that would work. The question is of course is should we keep the actual startup API and document the above workaround in the troubleshooting section?

Also the actual API can't return tasks for middleware-started sagas

@slorber 
>I guess the idea is that the sagas should be able to become "stable" before receiving any new dispatch from outside the sagas right? 

Yes in some sense. Actually I considered the global lock for all the effects, but afterthought it wasn't necessary; the only way a saga can affect another saga is by dispatching actions that the other saga will take. So i think that serializing put effects should be sufficient.

>However a nice default setting is cool, but I think it would be nice to give options in the put effect so that developers can decide themselves if they need sync/async/default for their puts

Yes I think it's a good idea.

**Comment by ronanwize** [comment]
I'll throw in my 2 cents: Go ahead and do it asynchronously (all our code already assumes that anyway, to make the sagas easier to reason about), and make the breaking change to satisfy the need to run actions at startup. We initialize our middleware in one place so it would be a very minor change.

**Comment by yelouafi** [comment]
see 0.10.0

**Comment by phaistonian** [comment]
Oh boy.

A lot to grasp and refactor on Monday then :-)
